src/Layers/xrRenderPC_GL/r2_R_sun.cpp

```diff
@@ -41,33 +42,58 @@ static int facetable[6][4] = {
 #define ALMOST_ZERO(F) ((FLT_AS_DW(F) & 0x7f800000L)==0)
 #define IS_SPECIAL(F) ((FLT_AS_DW(F) & 0x7f800000L)==0x7f800000L)
 
+void XRVec3TransformCoordArray(glm::vec3* out, const glm::vec3* in, const glm::mat4& matrix, unsigned int elements)
+{
+    for (unsigned int i = 0; i < elements; ++i)
+        out[i] = glm::vec3(glm::translate(matrix, in[i]) * glm::vec4(1.f,1.f,1.f,1.f));
+}
+
+void XRMatrixOrthoOffCenterLH(Fmatrix *pout, float l, float r, float b, float t, float zn, float zf)
+{
+    pout->identity();
+    pout->m[0][0] = 2.0f / (r - l);
+    pout->m[1][1] = 2.0f / (t - b);
+    pout->m[2][2] = 1.0f / (zf -zn);
+    pout->m[3][0] = -1.0f -2.0f *l / (r - l);
+    pout->m[3][1] = 1.0f + 2.0f * t / (b - t);
+    pout->m[3][2] = zn / (zn -zf);
+}
+
+void XRMatrixInverse(Fmatrix *pout, float *pdeterminant, const Fmatrix &pm)
+{
+    glm::mat4 out = glm::inverse(glm::make_mat4x4(&pm.m[0][0]));
+    
+    *pout = *(Fmatrix*)glm::value_ptr(out);
+    return;
+}
+
```

Comment:
Why math code is defined and implemented in render logic? Maybe better to implement it in separate file ( in xrCore/Math for example ), because this is just math helper functions. And i think better will be use separate namespace for all new math functions, for example xrMath::MatrixInverse. If we will change all D3DXMatrix functions in future to glm then that approach will be better i think.

Reply:
For general mathematics, there is already xrMiscMath, and D3DXMath was be used only in sun/thunder render because the engine mathematics is too slow. The simplest example - xrMiscMath does not use RVO and NRVO optimizations.

Reply:
Yes, but xrMiscMath doesn't have an MatrixInverse code. Inverse of matrix should be in general mathematics too, because its  just matrix functionality. Why we need tree different places with different math code. I understand that core of engine have own math functions, but if we adding something new i think we should put it in one place to be sure that we can find all in one place and to be able to use it dynamically in all places of code if we need.

Reply:
xrMiscMath already have unaccelerated functions:
```
Fmatrix& Fmatrix::invert(const Fmatrix& a);
bool     Fmatrix::invert_b(const Fmatrix& a);
Fmatrix& Fmatrix::invert_44(const Fmatrix& a);
```